pulp_container/app/tasks/sign.py

```diff
@@ -0,0 +1,128 @@
+import base64
+import hashlib
+import os
+import tempfile
+
+from django.conf import settings
+
+from pulpcore.plugin.models import Repository
+
+from pulp_container.app.models import (
+    ManifestSignature,
+    ManifestSigningService,
+    Tag,
+)
+
+from pulp_container.app.utils import extract_data_from_signature
+from pulp_container.constants import (
+    MANIFEST_MEDIA_TYPES,
+    SIGNATURE_TYPE,
+)
+
+
+def sign(repository_pk, signing_service_pk, reference, tags_list=None):
+    """
+    Create a new repository version by signing manifests.
+
+    Create signature for each manifest that is specified and add it to the repo.
+    If no manifests were specified, then sign all manifests in the repo.
+
+    What manifests to sign is identified by tag.
+    Manifest lists are not signed. Image manifests from manifest list are signed by digest.
+
+    Args:
+        repository_pk (uuid): A pk for a Repository for which a new Repository Version should be
+                             created.
+        signing_service_pk (uuid): A pk of the signing service to use.
+        reference (str): Reference that will be used to produce signature.
+        tags_list (list): List of PKs for :class:`~pulp_container.app.models.Tag` manifests of which
+                          should be signed.
+
+    """
+    repository = Repository.objects.get(pk=repository_pk).cast()
+    latest_version = repository.latest_version()
+    if tags_list:
+        latest_repo_content_tags = latest_version.content.filter(
+            pulp_type=Tag.get_pulp_type(), pk__in=tags_list
+        )
+    else:
+        latest_repo_content_tags = latest_version.content.filter(pulp_type=Tag.get_pulp_type())
+    latest_repo_tags = Tag.objects.filter(pk__in=latest_repo_content_tags)
+    signing_service = ManifestSigningService.objects.get(pk=signing_service_pk)
+    added_signatures = []
+    already_signed = []
+    for tag in latest_repo_tags:
+        tagged_manifest = tag.tagged_manifest
+        if tagged_manifest.media_type in MANIFEST_MEDIA_TYPES.IMAGE:
+            signature_pk = create_signature(
+                tagged_manifest, ":".join((reference, tag.name)), signing_service
+            )
+            added_signatures.append(signature_pk)
+        elif tagged_manifest.media_type in MANIFEST_MEDIA_TYPES.LIST:
+            # parse ML and sign per-arches by digest
+            for manifest in tagged_manifest.listed_manifests.iterator():
+                # image manifests can be present in multiple ML within the repo
```

Comment:
Is it possible, that a manifest can be part of a list and tagged directly? Also can multiple tags not point to the same manifest?

Reply:
yes, it is possible that manifest can be part of manifest list and tagged directly, in that case 2 signatures will be produced - one referenced by tag, second by digest.
and yes different tags can point to different manifests, but i am not sure i've understood your second question.

Reply:
I was wondering, if we need to extend the "i have singed this already" logic to all manifests we sign. But are you saying that the signature is different for two tags pointing at the same manifest? This is all a little bit confusing to me.

Reply:
if you have tag for and bar pointing to same manifest - yes it will produce 2 signatures, moreover if you trigger this task twice, second task will produce again 2 signatures because the signature contains the timestamp that changes the signature payload therefore the digest changes.

Reply:
I agree, it's confusing.
The signature here is actually a non-encrypted signed JSON. It is recommended for this JSON to have a tag specified if it's available. User can configure such a policy that a tag will be used in it. E.g. podman pull only with the tag `latest` and with the valid signature by produced by that key.